### PR TITLE
Add Doppler SAP Service

### PR DESCRIPTION
**Pull request by nreal**
# Background
Add a Doppler SAP service to send data from Doppler.Currency and Doppler database(Not implemented for the moment)

SAPService

Example Model to send to SAP service:
`[{"Entity":{"Date":"2020-03-27","SaleValue":65.5000,"BuyValue":60.5000,"CurrencyName":"Peso Argentino","CurrencyCode":"ARS"}}]`

### Change 1
Remove ILoggerAdapter and add HelperExtension for VerifyLogger in Unit tests

### Change 2
Change responsibility of CurrencyDto by CurrencyResponse

### Change 3
Add information in Hangfire UI to send to SAP System

- Empty list

emptyList

- ARS currency

ArsList

Could you review it?

### CODE REVIEW

File: Doppler.Sap.Job/DopplerSapJob.cs [hunk 1]
```diff
@@ -2,47 +2,54 @@
...
             var currencyDto = await _dopplerCurrencyService.GetCurrencyByCode();
-            
-            if (currencyDto.Any())
-            {
-                _logger.LogInformation($"Sending data to Sap system with data: {currencyDto}.");
-                //TODO: Add call DopplerSap service
-            }
+
+            if (!currencyDto.Any()) 
+                return "Not exist Currency for this date, please check if a holiday.";
```
Comment (fgchaio) [inline Doppler.Sap.Job/DopplerSapJob.cs]:
```suggestion
                return "Non-existent Currency for this date, please check if it's a holiday.";
```

**Comment by fgchaio** [comment]
great work! 👏

File: Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Net.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Doppler.Sap.Job.Service.DopplerCurrencyService
+{
+    public abstract class BaseExternalService
```
Comment (andresmoschini) [inline Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs]: This class has no responsibility, what do you think about removing it?

File: Doppler.Sap.Job/DopplerSapJob.cs [hunk 3]
```diff
@@ -2,47 +2,54 @@
...
-        public void Run()
+        public object Run()
         {
-            RunAsync().GetAwaiter().GetResult();
+            var a = RunAsync().GetAwaiter().GetResult();
+            return a;
         }
 
-        private async Task RunAsync()
+        private async Task<object> RunAsync()
```
Comment (andresmoschini) [inline Doppler.Sap.Job/DopplerSapJob.cs]: What is the meaning of this returned object?

File: Doppler.Jobs.Server/Startup.cs [hunk 4]
```diff
@@ -66,16 +64,26 @@ public void ConfigureServices(IServiceCollection services)
...
                 sp.GetService<IHttpClientFactory>(), 
                 httpClientPolicies,
                 dopplerCurrencySettings,
-                sp.GetService<ILoggerAdapter<DopplerCurrencyService>>(),
+                sp.GetService<ILogger<BaseExternalService>>(),
                 jobsConfig));
 
+            var dopplerSapServiceSettings = new DopplerSapServiceSettings();
+            Configuration.GetSection("DopplerSapService").Bind(dopplerSapServiceSettings);
+            services.AddSingleton(dopplerSapServiceSettings);
```
Comment (andresmoschini) [inline Doppler.Jobs.Server/Startup.cs]: If I am not wrong, you could use _IOptions_, maybe you can ask @VicThorX or @elsupergomez about it.

File: Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs (hunk 2 above)
Comment (nreal) [reply Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs]: Yes, is an abstract class that I used on another approach, It's unnecessary I'll remove

File: Doppler.Jobs.Server/Startup.cs [hunk 5]
```diff
@@ -71,7 +71,7 @@ public void ConfigureServices(IServiceCollection services)
                 sp.GetService<IHttpClientFactory>(), 
                 httpClientPolicies,
                 dopplerCurrencySettings,
-                sp.GetService<ILogger<BaseExternalService>>(),
+                sp.GetService<ILogger<DopplerCurrencyService>>(),
```
Comment (andresmoschini) [inline Doppler.Jobs.Server/Startup.cs]: Is it not possible simply declaring the implementation type like:

```csharp
            services.AddTransient<DopplerCurrencyService>()
```

Then, when you want to implemenit, the dependencies will be resolved directly.

File: Doppler.Sap.Job/DopplerSapJob.cs (hunk 1 above)
Comment (nreal) [reply Doppler.Sap.Job/DopplerSapJob.cs]: Done

File: Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs (hunk 2 above)
Comment (nreal) [reply Doppler.Sap.Job/DopplerCurrencyService/BaseExternalService.cs]: Done

File: Doppler.Sap.Job/DopplerSapJob.cs (hunk 3 above)
Comment (nreal) [reply Doppler.Sap.Job/DopplerSapJob.cs]: The meaning is that RunAsync() method return the object to displays in Hangfire UI the object created to send to DopplerSap
X7bwj


See [link](https://stackoverflow.com/questions/57748867/how-to-return-a-return-value-from-hangfire)

Comment (andresmoschini) [reply Doppler.Sap.Job/DopplerSapJob.cs]: 👍 

It is ugly, but it is Hangfire interface, so, I am fine. Thanks!

Comment (nreal) [reply Doppler.Sap.Job/DopplerSapJob.cs]: Maybe is possible to return an `IList<CurrencyResponse>`, let me check

Comment (andresmoschini) [reply Doppler.Sap.Job/DopplerSapJob.cs]: Nah, Object is fine

File: Doppler.Jobs.Server/Startup.cs (hunk 4 above)
Comment (nreal) [reply Doppler.Jobs.Server/Startup.cs]: Thanks, I resolved it in this commit https://github.com/MakingSense/doppler-jobs/pull/2/commits/24153e86346d2caf53371f24a469e6b5f355048b

Comment (andresmoschini) [reply Doppler.Jobs.Server/Startup.cs]: 👏

File: Doppler.Jobs.Server/Startup.cs (hunk 5 above)
Comment (nreal) [reply Doppler.Jobs.Server/Startup.cs]: Yes done https://github.com/MakingSense/doppler-jobs/pull/2/commits/932b523fac350bbf9be7fa7e2302d4b609a1650b

File: Doppler.Jobs.Server/Startup.cs [hunk 6]
```diff
@@ -74,15 +74,9 @@ public void ConfigureServices(IServiceCollection services)
                 sp.GetService<ILogger<DopplerCurrencyService>>(),
                 jobsConfig));
 
-            var dopplerSapServiceSettings = new DopplerSapServiceSettings();
-            Configuration.GetSection("DopplerSapService").Bind(dopplerSapServiceSettings);
-            services.AddSingleton(dopplerSapServiceSettings);
+            services.Configure<DopplerSapServiceSettings>(Configuration.GetSection("DopplerSapService"));
```
Comment (elsupergomez) [inline Doppler.Jobs.Server/Startup.cs]:
```suggestion
            services.Configure<DopplerSapServiceSettings>(Configuration.GetSection(nameof(DopplerSapServiceSettings)));
```

File: Doppler.Jobs.Server/Startup.cs [hunk 7]
```diff
@@ -66,15 +64,15 @@ public void ConfigureServices(IServiceCollection services)
...
             var jobsConfig = new TimeZoneJobConfigurations
             {
-                TimeZoneJobs = $"{Configuration["TimeZoneJobs"]}"
+                TimeZoneJobs = Configuration["TimeZoneJobs"]
             };
+            services.AddSingleton(jobsConfig);
+
+            services.AddTransient<DopplerCurrencyService>();
+
+            services.Configure<DopplerSapServiceSettings>(Configuration.GetSection("DopplerSapService"));
```
Comment (VicThorX) [inline Doppler.Jobs.Server/Startup.cs]: 👏

File: Doppler.Jobs.Server/Startup.cs (hunk 6 above)
Comment (nreal) [reply Doppler.Jobs.Server/Startup.cs]: Done in this commit https://github.com/MakingSense/doppler-jobs/pull/2/commits/ad517b6d93d2cbe37742b20316b203812b511c8b

**Pull request merged**